lms/models/lti_user.py

```diff
@@ -19,6 +21,11 @@ class LTIUser(NamedTuple):
     display_name: str
     """The user's display name."""
 
+    @property
+    def h_user(self):
+        """Return a models.HUser generated from this LTIUser."""
+        return HUser.from_lti_user(self)
```

Comment:
If the above `username()` function were public this would be something like:
```
return HUser(
    username=username(
        self.tool_consumer_instance_guid, self.lti_user.user_id
    ),
    display_name=self.display_name,
    provider=self.tool_consumer_instance_guid,
    provider_unique_id=self.lti_user.user_id,
)
```

Reply:
Hmm. I'm not sure I prefer this. I don't like it because:

1. It's dividing up the knowledge of how to construct an `HUser` from an `LTIUser` across two different places. There's a `models.h_user.h_username()` function that knows how to create the username from parts of an `LTIUser`, then `models.LTIUser.h_user`, in a different files, has the _rest_ of the knowledge about what parts of an `LTIUser` line up with what parts of the `HUser`.

2. It's encoding a bunch of h-specific knowledge (specifically: knowledge about how we create h users with valid `provider`'s and `provider_unique_id`'s etc) in the `LTIUser` class, which seems like the wrong place for it. Specifically, it means the `LTIUser` class knows that we use an `LTIUser`'s `tool_consumer_instance_guid` for the `HUser`'s `provider` and knows that we use an `LTIUser`'s `user_id` for the `HUser`s `provider_unique_id`. But `provider` and `provider_unique_id` are h-specific concepts that have nothing to do with LTI users. Shouldn't the `HUser` class be the class that knows about how to create a valid `HUser`?

   When all `LTIUser` has to do is `HUser.from_lti_user(self)`, then `LTIUser` is freed of any specific knowledge about what parts an `HUser` has or how they're constructed

```python
diff --git a/lms/models/h_user.py b/lms/models/h_user.py
index 5975aade..957316d4 100644
--- a/lms/models/h_user.py
+++ b/lms/models/h_user.py
@@ -30,21 +30,9 @@ class HUser(NamedTuple):
     def userid(self, authority):
         return f"acct:{self.username}@{authority}"
 
-    @classmethod
-    def from_lti_user(cls, lti_user):
-        provider = lti_user.tool_consumer_instance_guid
-        provider_unique_id = lti_user.user_id
-
-        def username():
-            """Return the h username for the current request."""
-            username_hash_object = hashlib.sha1()
-            username_hash_object.update(provider.encode())
-            username_hash_object.update(provider_unique_id.encode())
-            return username_hash_object.hexdigest()[:30]
-
-        return cls(
-            username=username(),
-            display_name=lti_user.display_name,
-            provider=provider,
-            provider_unique_id=provider_unique_id,
-        )
+
+def h_username(lti_user):
+    username_hash_object = hashlib.sha1()
+    username_hash_object.update(lti_user.tool_consumer_instance_guid.encode())
+    username_hash_object.update(lti_user.user_id.encode())
+    return username_hash_object.hexdigest()[:30]
diff --git a/lms/models/lti_user.py b/lms/models/lti_user.py
index 24ae9091..cbf0978c 100644
--- a/lms/models/lti_user.py
+++ b/lms/models/lti_user.py
@@ -1,6 +1,6 @@
 from typing import NamedTuple
 
-from lms.models import HUser
+from lms.models.h_user import HUser, h_username
 
 
 class LTIUser(NamedTuple):
@@ -24,7 +24,12 @@ class LTIUser(NamedTuple):
     @property
     def h_user(self):
         """Return a models.HUser generated from this LTIUser."""
-        return HUser.from_lti_user(self)
+        return HUser(
+            username=h_username(self),
+            display_name=self.display_name,
+            provider=self.tool_consumer_instance_guid,
+            provider_unique_id=self.user_id,
+        )
 
     @property
     def is_instructor(self):
```